class Solution {
    Map<Integer,Integer> map=new HashMap<>();
    public int coinChange(int[] coins, int amount) {
        int minCoins=dfs(coins,amount);
        return minCoins==Integer.MAX_VALUE?-1 : minCoins;
    }
   private int dfs(int[] coins,int amount){
    if(amount==0) return 0;
    if(map.containsKey(amount)){
        return map.get(amount);
    }
     int result=Integer.MAX_VALUE;
     for(int coin:coins){
        if(amount-coin>=0){
            int temp=dfs(coins,amount-coin);
            if(temp!=Integer.MAX_VALUE){
                result=Math.min(result,temp+1);
            }
        }
     }
     map.put(amount,result);
     return result;
   } 
}
